tempesta_fw/lb/tfw_lb_http.c

```diff
@@ -0,0 +1,1091 @@
+	r = tfw_connection_new(sk, Conn_Srv, srv, tfw_lb_reconnect_in_bg);
+	if (r) {
+		TFW_ERR("can't create connection object: srv=%p sk=%p"
+			" sock=%p\n", srv, sk, sock);
+		sock_release(sock);
+		return NULL;
+	}
+
+	TFW_DBG("connected: sock=%p sk=%p state=%d\n", sock, sk, sk->sk_state);
+	TFW_DBG_ADDR("connected to backend server", &srv->addr);
+	return sock;
+}
+
+/**
+ * Establish all possible connections to all servers in the @group.
+ *
+ * Try to open as much as possible connections to all servers in the group.
+ * Restore closed connections.
+ */
+static int
+tfw_lb_group_connect(TfwLbGroup *group)
+{
+	int i;
+	int ret = 0;
+	TfwLbSrv *srv;
+	struct socket *sock, *new_sock;
+
+	list_for_each_entry(srv, &group->srvs, list) {
+		for (i = 0; i < srv->socks_n; ++i) {
+			sock = srv->socks[i];
+
+			/* Skip healthy connections. */
+			if (sock && sock->sk->sk_state == TCP_ESTABLISHED)
+				continue;
+
+			/* Open a new connection. */
+			new_sock = tfw_lb_srv_connect(srv);
+			if (!new_sock) {
+				ret = -1;
+				break;
+			}
+
+			/* Replace/add the dead/nonexistent socket. */
+			/* TODO: update TfwLbGroup->sks here. */
+			rcu_assign_pointer(srv->socks[i], new_sock);
+			synchronize_rcu();
+			if (sock)
+				sock_release(sock);
+		}
+	}
+	return ret;
+}
+
+/**
+ * Close connections to all servers in the @group. Release all sockets.
+ */
+static void
+tfw_lb_group_disconnect(TfwLbGroup *group)
+{
+	TfwLbSrv *srv;
+	struct socket **sockp;
+
+	TFW_DBG("drain backend_group: %s\n", group->name);
+
+	list_for_each_entry(srv, &group->srvs, list) {
+		TFW_DBG_ADDR("disconnect backend", &srv->addr);
+
+		for (sockp = srv->socks; *sockp; ++sockp) {
+			TFW_DBG("release sock: %p\n", *sockp);
+			sock_release(*sockp);
+		}
+	}
+
+	memset(group->sks, 0, (group->sks_n * sizeof(struct sock *)));
+}
+
+static int
+tfw_lb_connect_all_groups(void)
+{
+	int r;
+	TfwLbGroup *group;
+
+	list_for_each_entry(group, &tfw_lb_groups, list) {
+		r |= tfw_lb_group_connect(group);
+	}
+
+	return r;
+}
+
+static void
+tfw_lb_disconnect_all_groups(void)
+{
+	TfwLbGroup *group;
+	list_for_each_entry(group, &tfw_lb_groups, list) {
+		tfw_lb_group_disconnect(group);
+	}
+}
+
+/**
+ * tfw_lb_connd() - The main loop of the tfw_lb_connd thread.
+ *
+ * The thread establishes connections to back-end servers in the background.
+ *
+ * It iterates over the list of all known servers and checks their connection
+ * pools, and if there are closed connections, the thread re-establishes them.
+ *
+ * When all connections are established, the thread sleeps until some connection
+ * is lost. It is woken by a callback which is called when connection is closed.
+ * Then the thread iterates over all servers as usual and reestablishes the
+ * closed connection.
+ *
+ * If a connection attempt is failed, instead of sleeping forever, it wakes up
+ * periodically and makes another attempt (with unlimited number of retries).
+ */
+static int
+tfw_lb_connd_mainloop(void *data)
```

Comment:
Not for this release. This depends on #83 and should be reworked when ss_connect() is implemented.